api/v1/storageprofile_types.go

```diff
@@ -0,0 +1,110 @@
+// +kubebuilder:validation:XValidation:rule="!has(oldSelf.blockPoolConfiguration) || has(self.blockPoolConfiguration)", message="blockPoolConfiguration is required once set"
+type StorageProfileSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="DeviceClass is immutable"
+	// +kubebuilder:validation:MaxLength=512
+	// DeviceClass is the deviceclass name.
+	DeviceClass string `json:"deviceClass"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="SharedFilesystemConfiguration is immutable"
+	// configurations to use for cephfilesystem.
+	SharedFilesystemConfiguration SharedFilesystemConfigurationSpec `json:"sharedFilesystemConfiguration,omitempty"`
+
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="BlockPoolConfiguration is immutable"
+	// configurations to use for  profile specific blockpool.
+	BlockPoolConfiguration BlockPoolConfigurationSpec `json:"blockPoolConfiguration,omitempty"`
+}
+
+// StorageProfileStatus defines the observed state of StorageProfile
+type StorageProfileStatus struct {
+	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// Phase describes the Phase of StorageProfile
+	// This is used by OLM UI to provide status information
+	// to the user
+	Phase string `json:"phase,omitempty"`
+}
+
+//+kubebuilder:object:root=true
+//+kubebuilder:subresource:status
+
+// StorageProfile is the Schema for the storageprofiles API
+type StorageProfile struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   StorageProfileSpec   `json:"spec,omitempty"`
+	Status StorageProfileStatus `json:"status,omitempty"`
+}
+
+//+kubebuilder:object:root=true
+
+// StorageProfileList contains a list of StorageProfile
+type StorageProfileList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           map[string]*StorageProfile `json:"items"`
+}
+
+// SPText stores a StorageProfile reconciliation message or status text
+type SPText int64
+
+// List of constants to show different reconciliation messages and statuses.
+const (
+	SPReconcileFailed SPText = iota
+	SPInit
+	SPReconcileCompleted
+	SPReconcileCompletedMessage
+)
+
+func (s SPText) String() string {
+	switch s {
+	case SPReconcileFailed:
+		return "StorageProfileReconcileFailed"
+	case SPInit:
+		return "StorageProfileInit"
+	case SPReconcileCompleted:
+		return "StorageProfileReconcileCompleted"
+	case SPReconcileCompletedMessage:
+		return "StorageProfile reconcile completed successfully"
+	}
+	return "unknown"
+}
```

Comment:
I can see that there are two concerns coupled here. A Phase value and some text messages. To me, it does not make sense to treat both the same. Messages are strings by nature, while phase can be argued to be an enum (but can be a string as well). Let's keep the 2 concerns separate.

A couple of other remarks:
-  We need to align with the established terminology for Phases. So "Ready" instead of "Completed"
- The actual string values for phases should not include the word `Reconcile` in them (the enum token can) 
- Unknown is not a phase, if you need to keep the field empty just keep it empty
- No reason to have a message that only appears once in the code as a cost (especially as it is a one-time message, not a predefined value)
- For outside facing value (like the status of CR) we should use fall names and not abbreviations or short names, so the text `init` should change to `initializing`
- switch cases have a default case, no need for an explicit return 
- With auto-complete in all of our editor available for us there is no need to got with shorting names, it just make the code less readable (SP prefix -> StorageProfile prefix) 
- Why does an external cluster have to do with storage profile phase/status?

```
const (
	StorageProfilePhase_Init = iota
	StorageProfilePhase_Failed      
	StorageProfilePhase_Ready
)

func (s SCText) StorageProfilePhaseToString() string {
	switch s {
	case StorageProfilePhaseInit:
		return "Initializing"
	case StorageProfilePhaseFailed:
		return "Failed"
	case StorageProfilePhaseReady:
		return "Ready"
	default:
	        return ""
	}
}
```

It also might be that a string subtype will work better than an enum as it doesn't need a conversion from int to string and is still different than just a plain old string and can be defined in the `StorageProfileTypes.go` file as the type of the phase field.

```
// StorageProfilePhase specifies a phase a storage profile reconciliation 
type StorageProfileString string

const (
	StorageProfilePhase_Init  StorageProfileString = "Initializing`
	StorageProfilePhase_Failed  StorageProfileString = Failed
	StorageProfilePhase_Ready StorageProfileString = Ready
)
```

Both approaches works for me

Reply:
With iota and switch I think you might need (un)marshal method as well or else it'll take base type 🤔.

Reply:
Refactored